시험 모드 및 시험 성적서 발급 기능 추가 및 일부 변경

ds5105119 opened the pull request:
### 추가한 기능

1. 시험 모드를 추가했습니다. 시험 모드에서는 문제 개수만 선택할 수 있으며, 15문제 이상을 선택해야 합니다.
2. start command를 사용 시 시험 모드를 켜고 끌 수 있게 바꿨습니다. 시험 모드를 실행하는 경우 ./ 위치에 "certification-%Y%m%d%H%M%S.png" 형식의 파일이 생성됩니다.

---

### 수정 사항

orm_study/script/certification.py 파일 추가
- pillow 모듈 사용
- generate_certification_image 함수 추가: 시험 성적서 PNG 파일 생성 함수

orm_study/script/decorator.py 파일 추가
- TimeTrace Class Decorator 추가

orm_study/quiz/_resources 폴더 추가
- 인증서 이미지 certification.png 추가
- 인증서 이미지 출력에 사용될 폰트 PretendardVariable.ttf 추가

orm_study/quiz/command.py 파일 수정
- fetch_exam 함수 추가
- select_exam_mode 함수 추가
- fetch_quiz 함수 변경
- select_quiz_num 함수 변경: min_num 키워드가 추가되었습니다. 기본값은 1이며, min_num 이상의 값을 선택해야 합니다.
- solve_quiz 함수에 TimeTrace Decorator 추가 -> 반환 값: Tuple[List[bool], float]: solve_quiz의 반환값과 solve_quiz 함수의 실행 시간
- start 함수 변경: 시험 모드, 퀴즈 모드 선택 기능 추가, 시험 모드 선택 시 certification.py 내의 함수로 성적서 발급하는 로직 추가
- 몇몇 함수에 Google Style Python Docstrings 추가

orm_study/quiz/reader.py 파일 수정
- get_chapter 로직 수정

orm_study/quiz/python.py 파일 수정
- PythonQuizManager Class의 solve 함수에서 solve_quiz 함수를 사용하므로, 반환받을 값 process_time를 추가

orm_study/tests/test_quiz_command.py 파일 수정
- test_solve_quiz 함수에서 solve_quiz를 사용하므로, 반환받을 값 process_time을 추가

requirements.txt 파일 수정
- pillow~=10.4.0 추가

requirements-dev.txt 파일 수정
- pillow~=10.4.0 추가

ds5105119 committed:
add certification.py
-add package: pillow
-add function: generate_certification_image

modified command.py
-add class: TimeTrace: for timer decorator
-add function: fetch_exam: for exam
-add import: orm_study.quiz.certification
-change function: fetch_quiz
-change function: select_quiz_num: add min_num parameter
-change function: solve_quiz: decorate
-change function: start

modified reader.py
-add import: orm_study.quiz.constants.TYPE
-change function: get_chapter

modified constants.py
-add dict: TYPE

modified requirements.txt
-add pillow

modified TODO.md

ds5105119 committed:
add certification.py
-add package: pillow
-add function: generate_certification_image

modified command.py
-add class: TimeTrace: for timer decorator
-add function: fetch_exam: for exam
-add import: orm_study.quiz.certification
-change function: fetch_quiz
-change function: select_quiz_num: add min_num parameter
-change function: solve_quiz: decorate
-change function: start

modified reader.py
-add import: orm_study.quiz.constants.TYPE
-change function: get_chapter

modified constants.py
-add dict: TYPE

modified requirements.txt
-add pillow

modified TODO.md

ds5105119 committed:
add certification.py
-add package: pillow
-add function: generate_certification_image

modified command.py
-add class: TimeTrace: for timer decorator
-add function: fetch_exam: for exam
-add import: orm_study.quiz.certification
-change function: fetch_quiz
-change function: select_quiz_num: add min_num parameter
-change function: solve_quiz: decorate
-change function: start

modified reader.py
-add import: orm_study.quiz.constants.TYPE
-change function: get_chapter

modified constants.py
-add dict: TYPE

modified requirements.txt
-add pillow

modified TODO.md

ds5105119 committed:
add certification.py
-add package: pillow
-add function: generate_certification_image

modified command.py
-add class: TimeTrace: for timer decorator
-add function: fetch_exam: for exam
-add import: orm_study.quiz.certification
-change function: fetch_quiz
-change function: select_quiz_num: add min_num parameter
-change function: solve_quiz: decorate
-change function: start

modified reader.py
-add import: orm_study.quiz.constants.TYPE
-change function: get_chapter

modified constants.py
-add dict: TYPE

modified requirements.txt
-add pillow

modified TODO.md

ds5105119 committed:
add certification.py
-add package: pillow
-add function: generate_certification_image

modified command.py
-add class: TimeTrace: for timer decorator
-add function: fetch_exam: for exam
-add import: orm_study.quiz.certification
-change function: fetch_quiz
-change function: select_quiz_num: add min_num parameter
-change function: solve_quiz: decorate
-change function: start

modified reader.py
-add import: orm_study.quiz.constants.TYPE
-change function: get_chapter

modified constants.py
-add dict: TYPE

modified requirements.txt
-add pillow

modified TODO.md

ds5105119 committed:
Merge branch 'Jungminchae:main' into main

ds5105119 committed:
add certification.py
-add package: pillow
-add function: generate_certification_image

modified command.py
-add class: TimeTrace: for timer decorator
-add function: fetch_exam: for exam
-add import: orm_study.quiz.certification
-change function: fetch_quiz
-change function: select_quiz_num: add min_num parameter
-change function: solve_quiz: decorate
-change function: start

modified reader.py
-add import: orm_study.quiz.constants.TYPE
-change function: get_chapter

modified constants.py
-add dict: TYPE

modified requirements.txt
-add pillow

modified TODO.md

ds5105119 committed:
Merge remote-tracking branch 'origin/main'

ds5105119 committed:
add certification.py
-add package: pillow
-add function: generate_certification_image

modified command.py
-add class: TimeTrace: for timer decorator
-add function: fetch_exam: for exam
-add import: orm_study.quiz.certification
-change function: fetch_quiz
-change function: select_quiz_num: add min_num parameter
-change function: solve_quiz: decorate
-change function: start

modified reader.py
-add import: orm_study.quiz.constants.TYPE
-change function: get_chapter

modified constants.py
-add dict: TYPE

modified requirements.txt
-add pillow

modified TODO.md

ds5105119 committed:
add certification.py
-add package: pillow
-add function: generate_certification_image

modified command.py
-add class: TimeTrace: for timer decorator
-add function: fetch_exam: for exam
-add import: orm_study.quiz.certification
-change function: fetch_quiz
-change function: select_quiz_num: add min_num parameter
-change function: solve_quiz: decorate
-change function: start

modified reader.py
-add import: orm_study.quiz.constants.TYPE
-change function: get_chapter

modified constants.py
-add dict: TYPE

modified requirements.txt
-add pillow

modified TODO.md

ds5105119 committed:
add certification.py
-add package: pillow
-add function: generate_certification_image

modified command.py
-add class: TimeTrace: for timer decorator
-add function: fetch_exam: for exam
-add import: orm_study.quiz.certification
-change function: fetch_quiz
-change function: select_quiz_num: add min_num parameter
-change function: solve_quiz: decorate
-change function: start

modified reader.py
-add import: orm_study.quiz.constants.TYPE
-change function: get_chapter

modified constants.py
-add dict: TYPE

modified requirements.txt
-add pillow

modified TODO.md

Jungminchae commented:
- @ds5105119 살펴보고 코멘트 남기겠습니다. 
- 살펴보기 전에 테스트가 실패한 상태인데, 코드 포맷이 맞지 않나봅니다. pre-commit 이용해주시면 되겠습니다! .pre-commit-config.yaml 그대로 사용하면 됩니다!

ds5105119 commented:
> * @ds5105119 살펴보고 코멘트 남기겠습니다.
> * 살펴보기 전에 테스트가 실패한 상태인데, 코드 포맷이 맞지 않나봅니다. pre-commit 이용해주시면 되겠습니다! .pre-commit-config.yaml 그대로 사용하면 됩니다!

넵 확인했습니다! black과 pre-commit 설정하고 다시 올리겠습니다.

ds5105119 committed:
add certification.py
-add package: pillow
-add function: generate_certification_image

modified command.py
-add class: TimeTrace: for timer decorator
-add function: fetch_exam: for exam
-add import: orm_study.quiz.certification
-change function: fetch_quiz
-change function: select_quiz_num: add min_num parameter
-change function: solve_quiz: decorate
-change function: start

modified reader.py
-add import: orm_study.quiz.constants.TYPE
-change function: get_chapter

modified constants.py
-add dict: TYPE

modified requirements.txt
-add pillow

modified TODO.md

ds5105119 committed:
add certification.py
-add package: pillow
-add function: generate_certification_image

modified command.py
-add class: TimeTrace: for timer decorator
-add function: fetch_exam: for exam
-add import: orm_study.quiz.certification
-change function: fetch_quiz
-change function: select_quiz_num: add min_num parameter
-change function: solve_quiz: decorate
-change function: start

modified reader.py
-add import: orm_study.quiz.constants.TYPE
-change function: get_chapter

modified constants.py
-add dict: TYPE

modified requirements.txt
-add pillow

modified TODO.md

ds5105119 committed:
add certification.py
-add package: pillow
-add function: generate_certification_image

modified command.py
-add class: TimeTrace: for timer decorator
-add function: fetch_exam: for exam
-add import: orm_study.quiz.certification
-change function: fetch_quiz
-change function: select_quiz_num: add min_num parameter
-change function: solve_quiz: decorate
-change function: start

modified reader.py
-add import: orm_study.quiz.constants.TYPE
-change function: get_chapter

modified constants.py
-add dict: TYPE

modified requirements.txt
-add pillow

modified TODO.md

ds5105119 commented:
감사합니다! pre-commit run --all-files 한 뒤 모두 패스했습니다.

<details>
trim trailing whitespace.................................................Passed<br>
fix end of files.........................................................Passed<br>
check yaml...............................................................Passed<br>
check for added large files..............................................Passed<br>
black....................................................................Passed<br>
flake8...................................................................Passed<br>
</details>

ds5105119 committed:
add certification.py
-add package: pillow
-add function: generate_certification_image

modified command.py
-add class: TimeTrace: for timer decorator
-add function: fetch_exam: for exam
-add import: orm_study.quiz.certification
-change function: fetch_quiz
-change function: select_quiz_num: add min_num parameter
-change function: solve_quiz: decorate
-change function: start

modified reader.py
-add import: orm_study.quiz.constants.TYPE
-change function: get_chapter

modified constants.py
-add dict: TYPE

modified requirements.txt
-add pillow

modified TODO.md

ds5105119 commented:
f-string 오류 수정하였습니다.

ds5105119 committed:
Create pylint.yml

ds5105119 committed:
Delete .github/workflows/pylint.yml

ds5105119 committed:
add certification.py
-add package: pillow
-add function: generate_certification_image

modified command.py
-add class: TimeTrace: for timer decorator
-add function: fetch_exam: for exam
-add import: orm_study.quiz.certification
-change function: fetch_quiz
-change function: select_quiz_num: add min_num parameter
-change function: solve_quiz: decorate
-change function: start

modified reader.py
-add import: orm_study.quiz.constants.TYPE
-change function: get_chapter

modified constants.py
-add dict: TYPE

modified requirements.txt
-add pillow

modified TODO.md

ds5105119 committed:
Merge remote-tracking branch 'origin/main'

Jungminchae commented on scripts/certification.py:
```diff
@@ -0,0 +1,32 @@
+import datetime
```
library를 빌드할 때 orm_study안에 있는 코드들만 빌드가 됩니다. 따라서 certification.py와 decorator.py는 에러가 발생합니다.
scripts는 라이브러리 외적으로 테스트나 검증할 때 사용하는 스크립트를 넣습니다. 
orm_study/quiz 안으로 옮기면 좋을것 같습니다. certification.py, decorator.py 모두 옮겨주세요

--- 
이미지까지 추가하여 점수표를 만드신 것 센스 GOOD!
추가된 코드에 대해서 전반적으로 가독성 좋고 깔끔한 코드 같습니다.

ds5105119 replied:
certification.py decorator.py 파일을 scripts 폴더에서 quiz 폴더로 이동하였습니다.

Jungminchae commented on orm_study/quiz/reader.py:
```diff
@@ -6,8 +6,18 @@
 def get_chapter(chapter_number: str, _type="1") -> str | tuple[str, str]:
```
이렇게 하는게 더 명시적이어서 좋아보입니다 👍

Jungminchae commented on tests/test_quiz_command.py:
```diff
@@ -59,7 +59,7 @@ def test_solve_quiz(monkeypatch):
     monkeypatch.setattr("rich.prompt.Prompt.ask", lambda prompt: next(inputs))
```
- 기존 코드를 수정한 이후에는 `pytest` 명령으로 테스트를 실행해보세요. → 테스트코드도 변경할 필요가 있으면 수정해야합니다. → 귀찮기는한데 보다 견고하게 프로젝트를 관리하기 위함 입니다. 
---
- `test start`는 시작했을 때 앞에 질문이 하나 더 생겨서 테스트가 실패합니다.
-  기존 테스트는 시험모드가 아닐 때를 테스트 하는 것이니 "a"를 추가하면 좋을 것 같습니다 ask도 하나 더 추가해야합니다.
def test_start(monkeypatch):
    inputs = iter(["a","1", "1", "1", "Answer 1"])
    monkeypatch.setattr("rich.prompt.Prompt.ask", lambda prompt, choices=None: next(inputs))
    monkeypatch.setattr("rich.prompt.Prompt.ask", lambda prompt, choices=None: next(inputs)) 
---
- `test_select_quiz_type_valid`, `test_select_quiz_type_invalid` 는 기존 코드에 choices가 빠져서 실패합니다
- 이렇게 바꾸면 될 것 같습니다.
def test_select_quiz_type_valid(monkeypatch):
      monkeypatch.setattr("rich.prompt.Prompt.ask", lambda prompt: "1")
      assert select_quiz_type() == "1"
---
- `test_fetch_quiz` 는  mock이 깨져서 실패가 났는데  `def mock_get_chapter(chapter_number: str, _type: str):` 이렇게 바꿔주면 될 것 같습니다.
---
테스트 작성에 익숙하지 않으실 것 같아서 다소 세세하게 적었는데 추후에 같이 테스트도 작성해보고 관련해서 같이 더 논의해보면 좋을 것 같습니다.

Jungminchae commented on orm_study/quiz/constants.py:
```diff
@@ -4,6 +4,16 @@
     "3": "연산자",
 }
 
+TYPE = {"1": "객관식", "2": "주관식", "3": "혼합"}
```
- TYPE 을 상수로 뺀 것도 좋습니다. 기존 코드처럼 하드코딩하는 습관은 좋지 않은데 그부분을 수정하신 것 GOOD

Jungminchae commented on orm_study/quiz/command.py:
```diff
@@ -1,62 +1,106 @@
+def select_quiz_num(max_num: int, min_num: int = 1) -> int:
     while True:
-        quiz_num = Prompt.ask(f"[white bold]문항 수를 선택해주세요. (최대 {max_num}개)")
+        msg = f"[white bold]문항 수를 선택해주세요. ({min_num} ~ {max_num}개)"
```
이런 방식으로 msg를 따로 뺀 것 아주 좋습니다. 가독성이나 유지보수성 면에서 훨씬 좋아요
```python
msg = f"[white bold]문항 수를 선택해주세요. ({min_num} ~ {max_num}개)"
quiz_num = Prompt.ask(msg)
```

Jungminchae commented on orm_study/quiz/command.py:
```diff
@@ -1,62 +1,106 @@
 
 def select_quiz_type() -> str:
     print(Panel(TYPE_CHOICE, title="문제 유형 선택", title_align="center"))
+
```
- 기존에 작성된 코드를 수정할 떄는 테스트가 깨질 수 있습니다. 
- 깨진다고 해서 문제되는 것은 아니고 변경된 로직에 맞게 테스트 코드를 수정하면 됩니다.

Jungminchae commented on scripts/decorator.py:
```diff
@@ -0,0 +1,21 @@
+    :param **kwargs: func's keyword arguments
+    :return: func's return value, func's executed time
+    """
+
```
- TimeTrace데코레이터 만드신거 굿입니다.
- Type hint는 Typer라이브러리와 관련된 코드가 아니면 사실 안써도 무방합니다만 쓴다면 통일감있게 다쓰면 좋을 것 같습니다.  
`def __init__(self, func: Callable[..., Any]) -> None:`

ds5105119 replied:
Type hint를 작성하였습니다.

Jungminchae commented on orm_study/quiz/command.py:
```diff
@@ -1,62 +1,106 @@
     return quiz_answer_set
 
 
+def select_exam_mode() -> bool:
```
- 새로 작성한 코드들에도 테스트코드가 있으면 더 좋을 것 같습니다!

ds5105119 replied:
새로 작성한 함수 select_exam_mode, fetch_exam에 대해 테스트 코드를 추가하였습니다.

Jungminchae reviewed (commented):
- 전반적으로 굉장히 잘해주셨습니다.
- Docstring까지 잘써주셔서 너무 좋았습니다. 다만 논의해보고 싶은점은 이 라이브러리 특성상 사용자가 완전 한국인일 것 같아서, 전반적인 문서화를 한국어로 하는 것은 어떨까 하는 생각인데 어떻게 생각하시는지요?!
- 테스트 실패하는 케이스들과 라이브러리 빌드할 때 문제되는 부분들이 있어서 리뷰남겨놨습니다.
- 수정할 때 참고해주시면 감사하겠습니다!

ds5105119 committed:
modified test_quiz_command.py
- change import
- add function: test_select_exam_mode_on
- add function: test_select_exam_mode_off
- add function: test_fetch_exam
- change function: test_select_chapter_invalid
- change function: test_select_quiz_type_valid
- change function: test_select_quiz_type_invalid
- change function: test_fetch_quiz
- change function: test_start

modified command.py
- change import
- add docstring to function: select_quiz_num

modified constants.py
- add CERTIFICATION_TTF_PATH
- add CERTIFICATION_IMG_PATH

move decorator.py from scripts to quiz
- change docstring from class: TimeTrace
- add type hint to function: TimeTrace.__init__

move certification.py form scripts to quiz
- add import
- add docstring to function: generate_certification_image
- change function: _load_font
- change function: generate_certification_image

ds5105119 committed:
modified test_quiz_command.py
- change import
- add function: test_select_exam_mode_on
- add function: test_select_exam_mode_off
- add function: test_fetch_quiz
- change function: test_select_chapter_invalid
- change function: test_select_quiz_type_valid
- change function: test_select_quiz_type_invalid
- change function: test_fetch_quiz
- change function: test_start

modified command.py
- change import
- add docstring to function: select_quiz_num

modified constants.py
- add CERTIFICATION_TTF_PATH
- add CERTIFICATION_IMG_PATH

move decorator.py from scripts to quiz
- change docstring from class: TimeTrace
- add type hint to function: TimeTrace.__init__

move certification.py form scripts to quiz
- add import
- add docstring to function: generate_certification_image
- change function: _load_font
- change function: generate_certification_image

ds5105119 committed:
Merge remote-tracking branch 'origin/main'

ds5105119 commented:
### 수정 사항

test_quiz_command.py
- command.py에 추가된 함수들에 대한 테스트 코드를 작성하였습니다.
- command.py에 수정된 함수들에 대한 테스트 코드를 수정하였습니다.
- 총 테스트 수는 13개이며, 13 테스트 중 13 테스트를 pass 하였습니다.

certification.py
- scripts 폴더에서 quiz 폴더로 이동되었습니다.
- constants.py의 CERTIFICATION_TTF_PATH, CERTIFICATION_IMG_PATH를 참조하도록 변경되었습니다. 각각 시험 성적서 발급을 위한 TTF 폰트 파일과 PNG 이미지 파일의 경로입니다.
- generate_certification_image 함수에서 사용하는 f-string 수정

decorator.py
- scripts 폴더에서 quiz 폴더로 이동되었습니다.
- TimeTrace class의 __init__ 함수에 대한 Type hint를 추가하였습니다.

그 이외의 Type hint 추가나 PEP 8: E501 관련 변경 사항이 있을 수 있습니다.

---

### 기타 사항
Docstring 관련
- 제가 보기에도 한글로 작성하는 것이 좋아보입니다.
- 이전에 추가된 Docstring을 전부 한글로 바꿨습니다. 논리적인 부분에서는 python 스타일로 작성하였습니다.

테스트 코드 작성을 도와주신 것에 대해 감사드립니다.

Jungminchae commented:
수고하셨습니다. GOOD
